### PR TITLE
Use lead image association throughout codebase

### PR DESCRIPTION
**Pull request by davidgisbey**
## Description

This moves over any code that uses `edition.images.first`  to determine the lead image to the new lead image association. The places this is found are:

- the case study presenter
- the uploaded images component
- the document summary page
- the lead_image model

It also updates the validation which stops publishers using the lead image within markdown. This was out of date and only covered the old syntax (`!!1`). I've updated this in 2 ways:

1. it now tests for the old syntax by using the index of the lead_image + 1. So if the lead image is the 3rd image it'll look for `!!3`
2. also matches against the new syntax (`[Image: {filename}]`

## Trello card

https://trello.com/c/5vCiIxeQ/1064-lead-image-handling-backend-work
 
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

### CODE REVIEW

File: test/unit/app/models/edition/custom_lead_image_test.rb [hunk 1]
```diff
@@ -4,28 +4,40 @@ class Edition::CustomLeadImageTest < ActiveSupport::TestCase
...
-    build(:draft_news_article, options)
-  end
+    edition = build(:draft_news_article, options)
+    image1 = build(:image)
+    image2 = build(:image)
 
-  test "reports that the first image is not available for adding inline" do
-    assert edition_with_custom_lead_image.image_disallowed_in_body_text?(1)
-  end
+    edition.stubs(:images).returns([image1, image2])
```
Comment (ryanb-gds) [inline test/unit/app/models/edition/custom_lead_image_test.rb:11]: What's the advantage of stubbing over setting up the association by doing `edition.images = [image1, image2]`? Or even creating the images first and passing them to the factory?

File: test/unit/app/models/edition/lead_image_test.rb [hunk 2]
```diff
@@ -2,41 +2,41 @@
 
 class Edition::LeadImageTest < ActiveSupport::TestCase
   test "should use placeholder image if none had been uploaded" do
-    model = stub("Target", images: [], lead_organisations: [], organisations: []).extend(Edition::LeadImage)
+    model = stub("Target", lead_image: nil, lead_organisations: [], organisations: []).extend(Edition::LeadImage)
```
Comment (ryanb-gds) [inline test/unit/app/models/edition/lead_image_test.rb:5]: I know you didn't write the original test, but do you know why this test uses stub everywhere instead of just building a model? I definitely think we should try to avoid that in favour of using factories with build stubbed if at all possible. Might be worth a separate PR for a refactor.

File: test/unit/app/models/speech_test.rb [hunk 3]
```diff
@@ -145,8 +145,9 @@ class SpeechTest < ActiveSupport::TestCase
   end
 
   test "search_index includes default image_url if it has one" do
-    image = create(:image)
-    speech = create(:published_speech, images: [image])
+    image = build(:image)
+    speech = build(:published_speech, document: build(:document))
```
Comment (ryanb-gds) [inline test/unit/app/models/speech_test.rb:149]: Does the below work? Again trying to avoid magic stubbing
```suggestion
    speech = build(:published_speech, document: build(:document), lead_image: image)
```

**Review by ryanb-gds — COMMENTED** [review]
A few comments, but generally looking good. Possibly a bit heavy use of stubbing in the test, but that might just be a taste thing

File: app/models/edition/custom_lead_image.rb [hunk 4]
```diff
@@ -4,14 +4,11 @@ module Edition::CustomLeadImage
   included do
     validates :body,
               format: {
-                without: /^!!1([^\d]|$)/,
-                message: "cannot have a reference to the first image in the text",
+                without: ->(edition) { /^!!#{edition.lead_image_index + 1}([^\d]|$)|\[Image: #{edition.lead_image.filename}\]/ },
```
[the remark this answers is not in the document]
Comment (ollietreend) [reply app/models/edition/custom_lead_image.rb:7]: Or alternatively, I wonder if it could be refactored into a custom validator method or class, rather than being defined as a 'format' rule?

That might make more 'room' to expand it into something more readable.

[the remark this answers is not in the document]
Comment (ollietreend) [reply app/models/edition/custom_lead_image.rb:7]: I should also point out that the `[Image:]` syntax allows any amount of whitespace (including none) before the filename.

For example, these are all equivalent:
- `[Image:example.jpg]`
- `[Image: example.jpg]`
- `[Image:    example.jpg]`

So it'd be good to have test coverage for this (and then the regex will need updating accordingly).

For what it's worth, [the regex used by Govspeak](https://github.com/alphagov/govspeak/blob/ae8136a9bba22ed55479974544fba27c2f98bd0a/lib/govspeak.rb#L356C24-L356C48) is:

```ruby
/^\[Image:\s*(.*?)\s*\]/
```

Where `\s` whitespace is allowed `*` times (zero or more).

Also, we should really make use of `Regex.escape()` for the image filename. Otherwise the `.` in the file extension will technically match _any character_, not just a literal `.` fullstop/period.

So something like:

```ruby
/\[Image: #{Regex.escape(edition.lead_image.filename)}\]/
```

File: test/unit/app/models/edition/lead_image_test.rb (hunk 2 above)
Comment (ollietreend) [reply test/unit/app/models/edition/lead_image_test.rb:5]: Maybe I'm too stub-averse, but I sort of feel the same about the stubbing done here, too:

https://github.com/alphagov/whitehall/blob/0005c564cbef8e74760c6c948eaf5a0449cdffbd/test/unit/app/models/edition/custom_lead_image_test.rb#L36-L37

It feels a bit brittle. What if we chain something else on to that Active Record call? e.g. `where()`

But presumably this is the 'cost' of using `build_stubbed` rather than `create` in these sorts of scenarios – e.g. when querying associations via the Active Record interface, that would usually expect data to have been persisted.

Comment (davidgisbey) [reply test/unit/app/models/edition/lead_image_test.rb:5]: yea unfortunately build stubbed isn't smart enough to handle active chaining active record queries. Generally if it's just one call like this i stub it out and any more i just use create. I just know we're trying to reduce our db hits.

Happy to use a create instead if thats the consensus.

File: test/unit/app/models/speech_test.rb (hunk 3 above)
Comment (davidgisbey) [reply test/unit/app/models/speech_test.rb]: thanks have sorted

File: test/unit/app/models/edition/custom_lead_image_test.rb (hunk 1 above)
Comment (davidgisbey) [reply test/unit/app/models/edition/custom_lead_image_test.rb]: thanks have sorted

File: app/components/admin/edition_images/uploaded_images_component.rb [hunk 5]
```diff
@@ -56,7 +59,7 @@ def all_image_asset_variants_uploaded?(image)
   def image_to_hash(image, index)
     {
       url: image.url,
-      preview_alt_text: index.zero? ? "Lead image" : "Image #{index}",
+      preview_alt_text: lead_image == image ? "Lead image" : "Image #{index}",
```
[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/components/admin/edition_images/uploaded_images_component.rb:60]: Yea agreed. I sort of half refactored this with the intention to come back after. Let me do a bit of work on that now and i'll push it up a fresh commit.

File: app/models/edition/custom_lead_image.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/models/edition/custom_lead_image.rb]: i've I've pulled the validation into it's own method now in this commit https://github.com/alphagov/whitehall/pull/8433/commits/71207de43df034120ee0f35ad9d657caad7269b lmn what you think

File: test/unit/app/models/edition/lead_image_test.rb (hunk 2 above)
Comment (davidgisbey) [reply test/unit/app/models/edition/lead_image_test.rb:5]: Regarding your first question, no i'm not actually sure. Happy to go back and refactor it to just use a case study or another model, but maybe in a follow up PR?

Comment (ryanb-gds) [reply test/unit/app/models/edition/lead_image_test.rb:5]: It definitely feels like we're fighting ActiveRecord a bit trying to stub database reads. I do see ActiveRecord as sacrificing testability for convenience to a large degree, we're so coupled to the database it's very hard to test our logic without it. I'd probably take the hit on the create for Ollie's example in favour of a simpler test, personally

Comment (davidgisbey) [reply test/unit/app/models/edition/lead_image_test.rb:5]: okay have updated

File: app/models/edition/custom_lead_image.rb [hunk 6]
```diff
@@ -37,4 +34,20 @@ def remove_lead_image
...
+      errors.add(:body, "cannot have a reference to the lead image in the text")
+    end
+  end
+
+  def body_contains_lead_image_filename_markdown?
+    body.match?(/\[Image:\s*#{Regexp.escape(lead_image.filename)}\s*\]/)
+  end
+
+  def body_contains_lead_image_index_markdown?
+    body.match?(/^!!#{lead_image_index + 1}([^\d]|$)/)
```
Comment (ollietreend) [inline app/models/edition/custom_lead_image.rb:51]: Could we try and see if this passes?

```suggestion
    body.match?(/!!#{lead_image_index + 1}[^\d]/)
```

File: app/models/edition/custom_lead_image.rb [hunk 7]
```diff
@@ -26,6 +17,12 @@ def update_lead_image
     edition_lead_image.save!
   end
 
+  def lead_image_index
+    return if lead_image.blank?
+
+    images.find_index(lead_image)
+  end
```
[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/models/edition/custom_lead_image.rb]: have sorted this

File: app/models/edition/custom_lead_image.rb (hunk 6 above)
Comment (davidgisbey) [reply app/models/edition/custom_lead_image.rb:47]: this doesn't work unfortunately. After some discussion we've decided to go as is

File: app/components/admin/edition_images/uploaded_images_component.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/components/admin/edition_images/uploaded_images_component.rb:60]: this was done in this PR https://github.com/alphagov/whitehall/pull/8440

File: app/models/edition/custom_lead_image.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/models/edition/custom_lead_image.rb]: I think we're okay on this front now but let me know if that's not the case

**Review by ollietreend — APPROVED** [review]
Thanks for doing so much hard work on this ⭐
